cmd/toolbox/main.go

```diff
@@ -0,0 +1,297 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/signal"
+	"strconv"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/urfave/cli/v2"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/chain/types"
+
+	"github.com/filecoin-project/curio/build"
+	"github.com/filecoin-project/curio/deps"
+)
+
+var log = logging.Logger("toolbox")
+
+func main() {
+	local := []*cli.Command{
+		precommitStuckCmd,
+	}
+
+	app := &cli.App{
+		Name:                 "toolbox",
+		Usage:                "Some tools to fix some problems",
+		Version:              build.UserVersion(),
+		EnableBashCompletion: true,
+		Commands:             local,
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:    "repo",
+				EnvVars: []string{"LOTUS_PATH"},
+				Hidden:  true,
+				Value:   "~/.lotus", // TODO: Consider XDG_DATA_HOME
+			},
+			&cli.StringFlag{
+				Name:  "log-level",
+				Value: "info",
+			},
+			&cli.StringFlag{
+				Name:     "actor",
+				Required: os.Getenv("LOTUS_DOCS_GENERATION") != "1",
+				Usage:    "miner actor to manage",
+				EnvVars:  []string{"SP_ADDRESS"},
+			},
+		},
+		Before: func(cctx *cli.Context) error {
+			return logging.SetLogLevel("toolbox", cctx.String("toolbox"))
+		},
+	}
+
+	// terminate early on ctrl+c
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-c
+		cancel()
+		fmt.Println("Received interrupt, shutting down... Press CTRL+C again to force shutdown")
+		<-c
+		fmt.Println("Forcing stop")
+		os.Exit(1)
+	}()
+
+	if err := app.RunContext(ctx, os.Args); err != nil {
+		log.Errorf("%+v", err)
+		os.Exit(1)
+		return
+	}
+
+}
+
+func SPTActorGetter(cctx *cli.Context) (address.Address, error) {
+	addr, err := address.NewFromString(cctx.String("actor"))
+	if err != nil {
+		return address.Undef, fmt.Errorf("parsing address: %w", err)
+	}
+	return addr, nil
+}
+
+var precommitStuckCmd = &cli.Command{
+	Name:  "precommit-stuck",
+	Usage: "Perform db operations to fix issues with precommit messages getting stuck",
+	Action: func(cctx *cli.Context) error {
```

Comment:
We should take a user input about which messages to check for. Doing a blanket try on all would be an issue. In fact, we should get a list of all pending ones. Then check if any of them can be found on filfox (already executed), check if FullNode is part the execution tipset. If it is then we show these messages to user and let me choose to fix the problem for specific ones they want. Or maybe have a flag like `--fix-all` then in that case fix them all.

Reply:
I agree we should create a pending(stuck)-msg cmd that lists msgs that are stuck. What changes would I need to make to db query to get the different stuck msgs? On the fixing other message type, can you give me an example of what other message/methods may get stuck in the same way as this? Currently, the fixing seems dependent on the precommit msg cid being the thing we are solving around.